graphql_compiler/compiler/ir_lowering_match/ir_lowering.py

```diff
@@ -8,27 +8,35 @@
 us to convert this Expression into other Expressions, using data already present in the IR,
 to simplify the final code generation step.
 """
+from typing import Dict, List, Optional, Set
+
 import six
 
-from ..blocks import Backtrack, CoerceType, MarkLocation, QueryRoot
+from ..blocks import Backtrack, CoerceType, Filter, MarkLocation, QueryRoot
+from ..compiler_entities import BasicBlock, Expression
 from ..expressions import BinaryComposition, FalseLiteral, Literal, TernaryConditional, TrueLiteral
+from ..helpers import Location
 from ..ir_lowering_common.location_renaming import (
     make_location_rewriter_visitor_fn,
     make_revisit_location_translations,
     translate_potential_location,
 )
-from .utils import convert_coerce_type_to_instanceof_filter
+from ..match_query import MatchQuery, MatchStep
+from ..metadata import QueryMetadataTable
+from .utils import CompoundMatchQuery, convert_coerce_type_to_instanceof_filter
 
 
 ##################################
 # Optimization / lowering passes #
 ##################################
 
 
-def rewrite_binary_composition_inside_ternary_conditional(ir_blocks):
+def rewrite_binary_composition_inside_ternary_conditional(
+    ir_blocks: List[BasicBlock],
```

Comment:
I have another tiny nit. Sorry for these tiny nits, but I think that might be relevant for future prs. 

Why not make this a Sequence instead?

Reply:
Do we pass `ir_blocks` as something other than a list elsewhere?

Reply:
That's a good point. I don't think we do. I would prefer type hints to be use the maximally general type. It makes the function more "powerful" and in this case, it would help make sure we weren't accidentally modifying the list. 

But I guess that this type hint is ok because it is consistent with the rest of the codebase. I'll approve the PR. Sorry for the troubles.

Meta-point:
Maybe we should follow the following convention when typing hinting a certain type:
1. If the type has never been type-hinted, then we should try to use the most general type.
2. If the type has been type-hinted before, we should either use the previously used type hint or change the existing type hint for a more general type.

Reply:
In general, the guidelines are sound. A few things come to mind, however, that I think make them remain guidelines rather than strict rules:
- Tools that do runtime analysis of code (e.g. `monkeytype` and `pyannotate`) will always suggest the most-specific type, since they don't and can't know which of the properties of the object are important (e.g. whether `Sequence` is okay or whether modification is important and it has to be `List`). Between more typing and less typing across the codebase, we should always prefer more, so we should in general make that type-hint-addition process as smooth as possible.
- It's difficult to know in general for existing code whether it is allowed some capability but chooses not to use it (e.g. mutating the list), versus whether it isn't allowed that capability and should only know `Sequence` instead. So when type-hinting existing code, it's not always clear whether `List` or `Sequence` is appropriate, nor is it feasible to always find that out (per the above point).

So while these guidelines are good for newly-designed and newly-written code, and are in the general good spirit of "be liberal in what you accept and strict in what you provide," I think applying them to type-hinting existing code must be done with care and good judgment.